include/swift/Frontend/Frontend.h

```diff
@@ -336,14 +337,31 @@ class CompilerInstance {
   enum : unsigned { NO_SUCH_BUFFER = ~0U };
   unsigned MainBufferID = NO_SUCH_BUFFER;
 
-  /// PrimaryBufferID corresponds to PrimaryInput.
-  unsigned PrimaryBufferID = NO_SUCH_BUFFER;
-  bool isWholeModuleCompilation() { return PrimaryBufferID == NO_SUCH_BUFFER; }
+  /// Identifies the set of input buffers in the SourceManager that are
+  /// considered primaries.
+  llvm::SetVector<unsigned> PrimaryBufferIDs;
 
-  SourceFile *PrimarySourceFile = nullptr;
+  /// Identifies the set of SourceFiles that are considered primaries. An
+  /// invariant is that any SourceFile in this set with an associated
+  /// buffer will also have its buffer ID in PrimaryBufferIDs.
+  llvm::SetVector<SourceFile*> PrimarySourceFiles;
+
+  /// Return whether there is an entry in PrimaryInputs for buffer \BufID.
+  bool bufferIDIsAPrimaryInput(unsigned BufID) const {
```

Comment:
Naming nitpicks: pulling the "is" up front makes it more obvious at the call site that this is a predicate: `isBufferIDAPrimaryInput`, `isBufferPrimaryInput` or even just `isPrimaryInput`.